go-chaos/cmd/cluster.go

```diff
@@ -0,0 +1,190 @@
+	"github.com/zeebe-io/zeebe-chaos/go-chaos/internal"
+)
+
+func AddClusterCommands(rootCmd *cobra.Command, flags *Flags) {
+	var clusterCommand = &cobra.Command{
+		Use:   "cluster",
+		Short: "Interact with the Cluster API",
+		Long:  "Can be used query cluster topology and to request dynamic scaling",
+	}
+	var statusCommand = &cobra.Command{
+		Use:   "status",
+		Short: "Queries the current cluster topology",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return printCurrentTopology(flags)
+		},
+	}
+	var waitCommand = &cobra.Command{
+		Use:   "wait",
+		Short: "Waits for a topology change to complete",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return waitForChange(flags)
+		},
+	}
+
+	rootCmd.AddCommand(clusterCommand)
+	clusterCommand.AddCommand(statusCommand)
+	clusterCommand.AddCommand(waitCommand)
+	waitCommand.Flags().IntVar(&flags.changeId, "changeId", 0, "The id of the change to wait for")
+}
+
+func printCurrentTopology(flags *Flags) error {
+	k8Client, err := createK8ClientWithFlags(flags)
+	if err != nil {
+		panic(err)
+	}
+
+	err = k8Client.AwaitReadiness()
+	if err != nil {
+		return err
+	}
+
+	port := 9600
+	closePortForward := k8Client.MustGatewayPortForward(port, port)
+	defer closePortForward()
+
+	topology, err := queryTopology(port)
+	if err != nil {
+		return err
+	}
+	formatted, err := json.MarshalIndent(topology, "", "  ")
+	if err != nil {
+		return err
+	}
+	fmt.Println(string(formatted))
+	return nil
+}
+
+func waitForChange(flags *Flags) error {
+	k8Client, err := createK8ClientWithFlags(flags)
+	if err != nil {
+		panic(err)
+	}
+
+	err = k8Client.AwaitReadiness()
+	if err != nil {
+		return err
+	}
+
+	port := 9600
+	closePortForward := k8Client.MustGatewayPortForward(port, port)
+	defer closePortForward()
+
+	for {
+		topology, err := queryTopology(port)
+		if err != nil {
+			return err
+		}
+		if topology.LastChange != nil && topology.LastChange.Id == int64(flags.changeId) {
+			if topology.LastChange.Status == "COMPLETED" {
+				internal.LogInfo("Change %d completed successfully", flags.changeId)
+				return nil
+			} else {
+				return fmt.Errorf("change %d failed with status %s", flags.changeId, topology.LastChange.Status)
+			}
+		} else if topology.LastChange != nil && topology.LastChange.Id > int64(flags.changeId) {
+			internal.LogInfo("Change %d is outdated but was most likely completed successfully, latest change is %d", flags.changeId, topology.LastChange.Id)
+			return nil
+		} else if topology.PendingChange != nil && topology.PendingChange.Id == int64(flags.changeId) {
+			competed := len(topology.PendingChange.Completed)
+			pending := len(topology.PendingChange.Pending)
+			total := competed + pending
+			internal.LogInfo("Change %d is %s with %d/%d operations complete", flags.changeId, topology.PendingChange.Status, competed, total)
+		} else {
+			internal.LogInfo("Change %d not yet started", flags.changeId)
+		}
+
+		time.Sleep(5 * time.Second)
```

Comment:
:wrench: I would add some maximum time out or something, this looks to me like it could potentially loop forever? I know we assume the change will eventually complete, but still... :sweat_smile:

Reply:
Yes we should definitely fail at some point.

Reply:
Yeah, good point :+1: Picking a good value is tricky though, I'd expect scaling to take anywhere from 5 minutes to >30 minutes..

Using that in an automated chaos test is problematic anyway because we'll run into job timeouts... :thinking:

Reply:
W00t 30 minutes 😆 😅

Reply:
I've set a timeout of 25 minutes now